### DISCUSSION TITLE
Provide alternative for `npm bin` use case

### DISCUSSION THREAD

**Discussion by ngraef** [npm/feedback#807] [CLI]
### Background

As of `npm@9`, [`npm bin` is removed](https://github.com/npm/statusboard/issues/537).

### Problem

My team uses some npm scripts that are structured like the following:

```json
"some-script": "node -r dotenv/config $(npm bin)/some-cli command arg"
```

The alternative to `npm bin` provided in the npm v9 release notes says "use `npx` or `npm exec` to execute binaries", though I don't believe they support this use case of injecting [`dotenv`](https://www.npmjs.com/package/dotenv).

### Alternatives

1. Hard code `node_modules/.bin/some-cli`: 👎 (this is bad because the path is an npm implementation detail, right?)
2. Use `$(which some-cli)`: I don't think this is portable to Windows
3. Is there a new npm command I don't know about that would solve this?
4. [`dotenv-cli`](https://www.npmjs.com/package/dotenv-cli) looks promising (`dotenv npx some-cli`?), though it's annoying to install another dependency for this

### Question
Is there any recommendation for updating this script to be compatible with npm v9?

**Comment by ljharb** [comment]
I’d love to see an `npm which` command which locates a binary using npm exec’s algorithm.

**Comment by wraithgar** [comment]
# This will not work in windows

Here's a band-aid if you do not need to support windows


```sh
~/D/n/s/npx $ cat package.json
{
  "name": "npx",
  "version": "1.0.0",
  "scripts": {
    "example": "node -r dotenv/config $(npm prefix -g)/bin/npx foo bar"
  },
  "dependencies": {
    "dotenv": "^16.0.3"
  },
  "bin": {
    "foo": "./foo.js"
  }
}
~/D/n/s/npx $ cat .env 
HELLO=WORLD
~/D/n/s/npx $ cat foo.js 
#!/usr/bin/env node

console.log(process.env.HELLO)
console.log(process.argv)
~/D/n/s/npx $ npm run example

> npx@1.0.0 example
> node -r dotenv/config $(npm prefix -g)/bin/npx foo bar

WORLD
[
  '/Users/wraithgar/.nvm/versions/node/v18.12.0/bin/node',
  '/Users/wraithgar/.npm/_npx/c95967b95d2bdfc5/node_modules/.bin/foo',
  'bar'
]

```

Unfortunately the "bin" directory in windows *is* the global prefix, not a subdirectory.

**Reply by ngraef** [reply]
Interesting option to take advantage of `npx`'s resolution. Thanks for the input.
